guillotina/contrib/redis/driver.py

```diff
@@ -23,22 +24,31 @@ def __init__(self):
         self._pubsub = None
         self._loop = None
         self._receivers = {}
+        self._pubsub_subscriptor = None
+        self._conn = None
         self.initialized = False
         self.init_lock = asyncio.Lock()
 
     async def initialize(self, loop):
         self._loop = loop
         async with self.init_lock:
             if self.initialized is False:
-                try:
-                    settings = app_settings["redis"]
-                    self._pool = await aioredis.create_pool(
-                        (settings["host"], settings["port"]), **settings["pool"], loop=loop
-                    )
-                    self._pubsub_subscriptor = aioredis.Redis(await self._pool.acquire())
-                    self.initialized = True
-                except AssertionError:
-                    logger.error("Error on initializing redis", exc_info=True)
+                while True:
+                    try:
+                        await self._connect()
+                        self.initialized = True
+                        break
+                    except Exception:  # pragma: no cover
+                        logger.error("Error initializing pubsub", exc_info=True)
+
+    @backoff.on_exception(backoff.expo, (OSError,), max_time=30, max_tries=4)
+    async def _connect(self):
+        settings = app_settings["redis"]
+        self._pool = await aioredis.create_pool(
+            (settings["host"], settings["port"]), **settings["pool"], loop=self._loop
+        )
+        self._conn = await self._pool.acquire()
+        self._pubsub_subscriptor = aioredis.Redis(self._conn)
```

Comment:
I think this changes doesn't fix what happend to us -- we restarted the redis server and guillotina wasn't able to recover. 

Looking at the code, I think that the problem is here. Since we use the same `self._conn` during the lifetime of the guillotina instance, if the connection dies doens't recover, since we don't return the connection to the pool (and can't be cleaned up). Therefore the pubsub_subscriptor.subscribre() keeps failing (as we experienced). Makes senes?

Reply:
Here is the trabeback of the error: https://github.com/plone/guillotina/issues/746#issue-531203035

Reply:
Yup, it makes sense. I just wasn't able to reproduce it by killing redis while guillotina was running and turning it back on. At least for, the connection would re-establish.

I'm going to play with it a bit more.

Reply:
I couldn't reproduce it locally as well...

Reply:
Check now @masipcat ?